src/lib/formatter_json.c

```diff
@@ -0,0 +1,115 @@
+/*
+ *   Software Updater - client side
+ *
+ *      Copyright © 2012-2019 Intel Corporation.
+ *
+ *   This program is free software: you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation, version 2 or later of the License.
+ *
+ *   This program is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU General Public License for more details.
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#define _GNU_SOURCE
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "formatter_json.h"
+#include "log.h"
+#include "progress.h"
+#include "strings.h"
+
+static int json_format = 0;
+
+static void log_json(FILE *out UNUSED, const char *file UNUSED, int line UNUSED, const char *label, const char *format, va_list args_list)
+{
+	json_message(label, format, args_list);
+}
+
+void set_json_format(void)
+{
+	json_format = 1;
+	log_set_function(log_json);
+	progress_set_format(json_progress, json_start, json_end);
+}
+
+void json_start(const char *op)
+{
+	if (json_format) {
+		fprintf(stderr, "[\n{ \"type\" : \"start\", \"section\" : \"%s\" },\n", op);
+		fflush(stderr);
+	}
+}
+
+void json_end(const char *op)
+{
+	if (json_format) {
+		fprintf(stderr, "{ \"type\" : \"end\", \"section\" : \"%s\" }\n]\n", op);
+		fflush(stderr);
+	}
+}
+
+void json_status(int status)
+{
+	if (json_format) {
+		fprintf(stderr, "{ \"type\" : \"status\", \"status\" : \"%d\" },\n", status);
```

Comment:
Use int instead of a string for status in the json output.

    status: 31

instead of

    status: "31"

I would move the status to the end json object:

    { type: "end", section: "bundle-add", status=31 }

Reply:
good idea about using int for numbers in the JSON output. I will change that also for the "progress" messages like this:
`{ "type" : "progress", "currentStep" : 1, "totalSteps" : 7, "stepCompletion" : 100, "stepDescription" : "download_bundle_manifests" }`

Reply:
about moving status to the `end` json object... don't you think there may be cases where we want to report an intermediate status during a process? so if we merge them then we won't have that flexibility. One possibility would be to include status in the `end` object but also leave the `json_status` function in case we want to use it some time for some use case. What do you think @otaviobp ?

Reply:
I don't know. Good point. Let it the way it is then

Reply:
I will add the status to the json `end` object, I anticipate this is the one we'll use most of the time, but I will also leave the `json_status` function in case we ever need it. We can remove it in the future if we don't see any use for it.

Reply:
great! I think use int instead string must be useful also for the gnome-software.